Tools/clinic/clinic.py

```diff
@@ -2055,7 +2056,16 @@ def write_file(filename: str, new_contents: str) -> None:
 ClassDict = dict[str, "Class"]
 DestinationDict = dict[str, Destination]
 ModuleDict = dict[str, "Module"]
-ParserDict = dict[str, "DSLParser"]
+
+
+class Parser(Protocol):
+
+    @abc.abstractmethod
+    def __init__(self, *args, **kwargs): ...
+
+    @abc.abstractmethod
+    def parse(self, block: Block) -> None: ...
```

Comment:
I would go for:

```suggestion
class Parser(Protocol):
    def __init__(self, clinic: Clinic) -> None: ...
    def parse(self, block: Block) -> None: ...
```

Several small things changed:
- The `__init__` method should be more precise, imo -- both `PythonParser` and `DSLParser` only take a single non-`self` argument, `clinic`. So let's mandate that in the interface we're specifying here!
- Protocols don't really need the `@abstractmethod` decorator if you're not inheriting from them, since all methods are implicitly abstract -- mypy won't recognise another class as a subtype of this protocol unless it implements all methods on the protocol. Having said that, you may prefer having `@abstractmethod` for extra explicitness, so I'm okay with it if you want to add it back :)
- (Style nit): I personally like to write protocol classes as if they're stubs, with no line break between the methods. This is just a personal preference, though; you can have line breaks between the methods if you prefer :)

Reply:
Thanks!